### PR TITLE
fix: predicate shall use real schema to create physical exprs

### PR DESCRIPTION
**Pull request by v0y4g3r**
I hereby agree to the terms of the [GreptimeDB CLA](https://gist.github.com/xtang/6378857777706e568c1949c7578592cc)

## What's changed and what's your intention?

This PR fixes the incorrect behavior when pruning memtable time series using the schema from `RegionMetadata`.

The bug originates from how `Predicate` used for pruning generates physical expr, because the indices for primary keys in primary key record batch may differ from the indices defined in `RegionMetadata`. Instead, in this fix we postpone the generation of physical exprs until evaluation and explicitly requires callers to provide a schema for physical expr generation so that those physical expr will have the save schema as record batch evaluated.

## Checklist

- [x]  I have written the necessary rustdoc comments.
- [x]  I have added the necessary unit tests and integration tests.

### CODE REVIEW

File: src/table/src/predicate.rs [hunk 1]
```diff
@@ -39,19 +40,24 @@ mod stats;
...
     /// evaluated against record batches.
     /// Returns error when failed to convert exprs.
-    pub fn try_new(exprs: Vec<Expr>, schema: SchemaRef) -> error::Result<Self> {
-        let arrow_schema = schema.arrow_schema();
-        let df_schema = arrow_schema
+    pub fn new(exprs: Vec<Expr>) -> Self {
+        Self { exprs }
+    }
+
+    #[inline]
```
[the remark this answers is not in the document]
Comment (v0y4g3r) [reply src/table/src/predicate.rs]: > 

fixed in https://github.com/GreptimeTeam/greptimedb/pull/2642/commits/8b0ad7e67fd03e619097fd934c31f31978b42fb1

File: tests/cases/standalone/common/select/prune.result [hunk 2]
```diff
@@ -0,0 +1,36 @@
+create table demo(ts timestamp time index, `value` double, host string,idc string, collector string, primary key(host, idc, collector));
+
+Affected Rows: 0
+
+insert into demo values(1,2,'test1', 'idc1', 'disk') ,(2,3,'test2', 'idc1', 'disk'), (3,4,'test3', 'idc2','memory');
+
+Affected Rows: 3
+
+select * from demo where host='test1';
```
[the remark this answers is not in the document]
Comment (v0y4g3r) [reply tests/cases/standalone/common/select/prune.result:9]: > Adds some tests for other tags such as `idc' and `collector`.

fixed in https://github.com/GreptimeTeam/greptimedb/pull/2642/commits/8b0ad7e67fd03e619097fd934c31f31978b42fb1

File: src/mito2/src/engine/prune_test.rs [hunk 3]
```diff
@@ -76,6 +76,16 @@ async fn test_read_parquet_stats() {
 +-------+---------+---------------------+
 | tag_0 | field_0 | ts                  |
 +-------+---------+---------------------+
+| 0     | 0.0     | 1970-01-01T00:00:00 |
```
[the remark this answers is not in the document]
Comment (v0y4g3r) [reply src/mito2/src/engine/prune_test.rs:79]: > The test looks too simple, can we add some tags?

Parquet pruning now only works for the first tag column, so adding more tags will not have any effect.

**Review by killme2008 — APPROVED** [review]
LGTM

**Review by waynexia — APPROVED** [review]
LGTM

File: src/table/src/predicate.rs [hunk 4]
```diff
@@ -123,7 +120,9 @@ impl Predicate {
 
     /// Prunes primary keys
     pub fn prune_primary_key(&self, primary_key: &RecordBatch) -> error::Result<bool> {
-        for expr in &self.exprs {
+        let pk_schema = primary_key.schema();
+        let physical_exprs = self.to_physical_exprs(&pk_schema)?;
```
[the remark this answers is not in the document]
Comment (v0y4g3r) [reply src/table/src/predicate.rs:124]: > question(non-blocking): Will this affect performance as we need to build the physical expression for each primary key?
> 
> We should run a benchmark after that. Maybe we need to define a PhysicalPredicate that holds the physical expression and reuse it in the memtable iter (or cache the physical expr in the iter).

This PR aims to minimize code change. We can cache the physical exprs in iterator for next usage.

**Pull request merged**